aerogear/unifiedpush-operator
# Provision UPS based on CR

JIRA: https://issues.jboss.org/browse/AEROGEAR-9009

~This doesn't yet create the Ingress or Route, since it's not 100%
clear what we need to do around that, and the interaction with the
oauth-proxy container in the UPS Pod.~

For now this just assumes OpenShift and creates a Route. This gives us something to iterate on, where we can have logic to check if OpenShift or plain K8s and act accordingly.

~Kubernetes doesn't appear to provide an OAuth server ootb like
OpenShift does, so the OpenShift oauth-proxy that we use may not work
there as easily.~

~Also, when setting the OAuthRedirectReference in the ServiceAccount,
the OpenShift docs currently claim that Route is the only supported
Kind at the moment.~

We might be able to use `oauth-redirecturi` for Ingress rather than `oauth-redirectreference`.

~There's also the more obvious differences between Route and Ingress:
we're used to using Route, where we just give it a name and the
fully-qualified URL gets provided to us automatically, and configured
in the Router. For k8s Ingress, we need to provide the host directly
in the Ingress resource, so for the operator to do that, it would need
to know what that should be: so it would probably need to be provided
by the person provisioning, in the UnifiedPushServer CR.~

### Verification steps
1. Checkout this branch
2. As a cluster admin user in OpenShift (because of role creation), run `make cluster/prepare`. This will create a `unifiedpush` project and other things needed for the operator.
3. Deploy the image built for this PR:

```
cat deploy/operator.yaml | sed -e "s|quay.io/aerogear/unifiedpush-operator|quay.io/aerogear/unifiedpush-operator:$(git rev-parse --short HEAD)|g" | kubectl create -f -
```

4. Check that the operator has started successfully with like `kubectl get po` or whatever, and have a look at the logs to make sure there's no obvious errors in there.

5. Create a UPS!

```
kubectl apply -f deploy/crds/aerogear_v1alpha1_unifiedpushserver_cr.yaml
```

6. All of the required resources should get created. ~It might happen that the UPS container itself doesn't get to a ready state, if the PostgreSQL wasn't ready in time for it. Unfortunately for now this means that you'll just have to delete the pod for it (I'll create follow-on issues to both wait for PostgreSQL, and also to add a liveness check to the UPS pod/container). Once it's deleted, a new one will be automatically created.~ I've added a liveness check here now in this PR, so it should get restarted until it works.

7. Once everything is ready, you should be able to login through the oauth-proxy in your browser:

```
xdg-open https://$(oc get route -o json | jq -r .items[0].spec.host)
```
